diff: Add support for anchoring specific lines in diff mode

ychin opened the pull request:
Adds support for anchoring specific lines to each other while viewing a diff. While lines are anchored, they are guaranteed to be aligned to each other in a diff view, allowing the user to control and inform the diff algorithm what the desired alignment is. Internally, this is done by splitting up the buffer at each anchor and run the diff algorithm on each split section separately, and then merge the results back for a logically consistent diff result.

To do this, add a new "diffanchors" option that takes a list of `{address}`, and a new "diffopt" option value "anchor". Each address specified will be an anchor, and the user can choose to use any type of address, including marks, line numbers, or pattern search. Anchors are sorted by line number in each file, and it's possible to have multiple anchors on the same line (this is useful when doing multi-buffer diff). Update documentation to provide examples.

This is similar to Git diff's `--anchored` flag. Other diff tools like Meld/Araxis Merge also have similar features (called "synchronization points" or "synchronization links"). We are not using Git/Xdiff's `--anchored` implementation here because it has a very limited API (it requires usage of the Patience algorithm, and can only anchor unique lines that are the same across both files).

Because the user could anchor anywhere, diff anchors could result in adjacent diff blocks (one block is directly touching another without a gap), if there is a change right above the anchor point. We don't want to merge these diff blocks because we want to line up the change at the anchor. Adjacent diff blocks were first allowed when linematch was added, but the existing code had a lot of branched paths where line-matched diff blocks were handled differently. As a part of this change, refactor them to have a more unified code path that is generalized enough to handle adjacent diff blocks correctly and without needing to carve in exceptions all over the place.

---

# Notes

As mentioned above a decent chunk of the change is the refactoring of the linematch code to remove as much `if (dp->is_linematched) { … }` special casing as possible as most of the time it duplicated functionality just to support adjacent diff blocks. The refactored done here is to simplify the code and make adjacent diff blocks a native concept rather than hidden in some if/else branching. Existing functionality should all still work. One caveat is that existing linematch code implicitly assumes `diffopt+=filler` to be set. The refactored code does not mandate that, but we now explicitly set filler to be true in code when linematch is set to match old behavior (and also because it doesn't make sense to use linematch without filler lines anyway).

This feature requires a bit of user action to take advantage of, so below are some examples to illustrate how it could be used.

## Examples

Some examples of actual commits that I have used diff anchors to help as I have been testing this locally for a few months:

### Example 1

In #16881 (commit 9943d479), there were a lot of code being moved around and refactored into functions. It's impossible for any diff algorithm to know which part of the code you are interested in. So if you try to make sense of what happened to the top-level function `diff_find_change` this is what you see in regular Vim diff:

<img width="893" alt="image" src="https://github.com/user-attachments/assets/3c597a73-30ed-4883-a15e-c4ec16872c8e" />

This is not ideal. However, if I re-anchor using `set diffopt+=anchor diffanchors=1/diff_find_change(/` I now get a much better sense of this particular spot in code:

<img width="891" alt="image" src="https://github.com/user-attachments/assets/e2bda4ad-7a2c-456b-91b7-d6c8f975f12a" />

If I'm interested how parts of this code got refactored into a child function (`diff_find_change_simple`) instead, I can simply re-anchor to now have them side-by-side (this can be done using marks and `set diffanchors='a`):

<img width="880" alt="image" src="https://github.com/user-attachments/assets/4263f99d-79bc-4089-81c8-6c4796e3bd2f" />

### Example 2

In #17579 (commit d75ab0c), I moved the text around in a help doc:

<img width="1138" alt="image" src="https://github.com/user-attachments/assets/75418a26-d571-4393-8d1b-44908612e776" />

The resulting diff is a little hard to make use of. So I could re-anchor the "There can be deleted…" lines and now the diff makes sense for this particular section so I could see if I introduced any typos:

<img width="1138" alt="image" src="https://github.com/user-attachments/assets/98ac1525-15be-4477-a638-f58f9a85c7bc" />

### Example 3

For this PR itself, the `diff_try_update()` function was quite difficult to see what's new versus old code due to indentation change. I could turn on `diffopt+=iwhite` to just ignore all whitespace, but another way was to just re-anchor quickly. Original diff:

<img width="1091" alt="image" src="https://github.com/user-attachments/assets/a8df2edd-1a5f-4f3b-8522-6373b70ff843" />

With the re-anchoring (by doing `set diffopt+=anchor dia='a,'b`). We can now still see the indentation change reflected in inline diff, but the existing code is now correctly shown as unchanged.

<img width="1091" alt="image" src="https://github.com/user-attachments/assets/01bee4b1-e605-4e04-8835-38c374056c27" />


### Multi-file example

Here is an example of how it could work in multi-file diff'ing. Below are three files where one function was moved, causing the diff results to be difficult to read (the marks are shown using a plugin):

<img width="925" alt="image" src="https://github.com/user-attachments/assets/69af5e49-0472-4ae5-b438-419f91649053" />

However, if we just use `set diffopt+=anchor diffanchors='a,'b`, we can realign things in a much nicer way. Note that in file 1, both marks are on the same line, which is allowed:

<img width="925" alt="image" src="https://github.com/user-attachments/assets/545db47d-f06f-46e2-88ac-1702d1216380" />

# FAQs

## Why not use Git / Xdiff's `--anchored` directly?

As mentioned above, Git's feature is quite limited. It is essentially doing the same thing (splitting the file up into multiple smaller sections), but it piggybacks on the patience algorithm and as such requires anchoring unique lines only. The new Vim diff anchors has no such restrictions and allows anchoring any lines that you can specify using a Vim `{address}`. It also works with any diff algorithm and also external diff tools.

## Doesn't line match already solve some of this?

Line match can solve a local alignment issue based on text similarity, but it cannot do this across the whole file. Also, this feature allows the user to manually place the anchors, which is important in ambiguous situations where there is no right answer.

## How does this relate to move detection?

Move detection is a feature in some diff tools that can detect moved code and highlights them. However, they don't always work when the moved code has some small changes, and they don't usually have a way to show that difference. Diff anchors is a more manual (as it requires user actions) but flexible method to look at moved code. It's still possible to implement move detection in Vim diff in the future though and they complement each other.

## Why do we need `diffopt+=anchor` when we already need to set `diffanchors` separately?

I think there could be times where you want to pre-set `diffanchors` but not immediately enable it. E.g. if you use anchors frequently you can do `set diffanchors='a,'b,'c` in vimrc and then just turn on/off anchors while viewing diff. That said it's possible to refactor this change to not have the extra `diffopt+=anchor` and just have it be implied from whether `diffanchors` is non-empty or not.

---

# TODOs

- [x] Implement tests.

ychin committed:
diff: Add support for anchoring specific lines in diff mode

Adds support for anchoring specific lines to each other while viewing a
diff. While lines are anchored, they are guaranteed to be aligned to
each other in a diff view, allowing the user to control and inform the
diff algorithm what the desired alignment is. Internally, this is done
by splitting up the buffer at each anchor and run the diff algorithm on
each split section separately, and then merge the results back for a
logically consistent diff result.

To do this, add a new "diffanchors" option that takes a list of
`{address}`, and a new "diffopt" option value "anchor". Each address
specified will be an anchor, and the user can choose to use any type of
address, including marks, line numbers, or pattern search. Anchors are
sorted by line number in each file, and it's possible to have multiple
anchors on the same line (this is useful when doing multi-buffer diff).
Update documentation to provide examples.

This is similar to Git diff's `--anchored` flag. Other diff tools like
Meld/Araxis Merge also have similar features (called "synchronization
points" or "synchronization links"). We are not using Git/Xdiff's
`--anchored` implementation here because it has a very limited API
(it requires usage of the Patience algorithm, and can only anchor
unique lines that are the same across both files).

Because the user could anchor anywhere, diff anchors could result in
adjacent diff blocks (one block is directly touching another without a
gap), if there is a change right above the anchor point. We don't want
to merge these diff blocks because we want to line up the change at the
anchor. Adjacent diff blocks were first allowed when linematch was
added, but the existing code had a lot of branched paths where
line-matched diff blocks were handled differently. As a part of this
change, refactor them to have a more unified code path that is
generalized enough to handle adjacent diff blocks correctly and without
needing to carve in exceptions all over the place.

habamax commented on runtime/doc/options.txt:
```diff
@@ -2950,6 +2950,20 @@ A jump table for the options with a short description can be found at |Q_op|.
+					    *'dia'* *'diffanchors'* *E1548*
+'diffanchors' 'dia'	string	(default "")
+			global or local to buffer |global-local|
+	List of {address} in each buffer, separated by commas, that are
```
What would be the meaning of the anchor if address is a current line `.` or the whole buffer `%`?

habamax replied:
I guess for the current line it would just anchor it and then after `:diffupdate` I should be able to see results.

habamax replied:
and the whole buffer probably wouldn't make sense to anchor.

ychin replied:
`.` would use the cursor location of the first diff window containing the buffer (there could be multiple windows for this buffer). It's not a terribly useful way to specify anchors, but it does work and has a consistent behavior. In `parse_diffanchors` we specifically set `curwin` and `curbuf` correctly to make sure these addresses correspond to the window with the buffer.

`%` would not work (setting it would trigger an error) because it's not a valid `{address}`. It's usually used as a `[range]` instead, as a shortcut for `1,$`. I decided not to add specialized handling for `%` because it would break how diff anchors are supposed to be specified (i.e. a comma-separated list of addresses) and I think the use cases of `%` for diff anchors is niche. FWIW, it's possible to add this, but if we want to add it, `%` would probably need to mean `1,$+1` for this use case.

ychin replied:
> and the whole buffer probably wouldn't make sense to anchor.

It could make sense to anchor the whole file sometimes, if set using buffer-local option. So if you set buffer 1's anchor to `setlocal dia=1,$+1`, and buffer 2 to say `setlocal dia='a,'b`, that means everything in buffer 1 will be diff'ed against only the part you specified in buffer 2. There *is* some use.

ychin replied:
> I guess for the current line it would just anchor it and then after :diffupdate I should be able to see results.

Yes. You could do that. Just move the cursor and call `:diffupdate`. You can do the same with setting marks, or selecting texts using visual mode (which updates the `<` / `>` marks). Even something like ``` `` ``` would work.

habamax replied:
> `%` would not work (setting it would trigger an error) because it's not a valid {address}

It is specified as equal to `1,$` but still within `:h {address}` though.

ychin replied:
> It is specified as equal to 1,$ but still within :h {address} though

Hm, I guess you are right. I would probably just specify in the 'diffanchors' docs that this is not supported than adding support for it, since it seems pretty low value for this feature. (Edit: Just updated the docs to reflect that)

ychin committed:
Update docs to be clearer on the constraints of the 'diffanchors' option

ychin committed:
Fix topfill calculation by clamping negative values

This could happen if we are setting diffopt multiple times, e.g. `set
diffopt=internal,filler diffopt+=anchor` which results in a stale
w_topfill that's not updated to reflect the new diff. The clamp makes
sure the calculated value is sane.

ychin committed:
Add a generic error when diff anchors fail to resolve at diff time

This gives a general error to let the user know the anchor resolution
has failed, in addition to whatever specific message they see like
"invalid mark". Also, currently a failed pattern search does not show
the error message since we are searching with silent mode on, so this is
the only error they will see.

ychin committed:
Make get_address() not add search pattern to @/ register

This prevents diff anchors using pattern search from polluting @/ and to
make sure it doesn't reset the search highlight. The only other user of
this function which relies on the silent flag is the incsearch but that
function already caches the pattern using `save_last_search_pattern()`.
We could do that for diff anchors too but it's simply more efficient to
just not pollute the register to begin with. It's unlikely a caller
passing "silent" in would want to do that.

ychin committed:
Add `dia` / `diffanchors` to Vimscript syntax files

ychin committed:
Docs minor typos

ychin committed:
Code clean ups

- Clean up prototype as we removed the `diff_check()` function

- Minor logic fix to remove unnecessary condition

ychin committed:
Add tests for diff anchors

This should cover all the new functionality as well as testing that
adjacent blocks still work properly regarding scrollbind and
diffget/diffput.

Add add checks to line match to test that the "filler" setting is done
implicitly whenever line match is used.

ychin committed:
Fix tags file not being sorted correctly

h-east commented on runtime/doc/diff.txt:
```diff
@@ -336,7 +337,129 @@ name or a part of a buffer name.  Examples:
+synchronize text across files.  Each anchor matches each other in each file,
+allowing you to control the output of a diff.
+
+This is useful when a change involves complicated edits. For example, if a
```
```suggestion
This is useful when a change involves complicated edits.  For example, if a
```

ychin replied:
Done

h-east commented on runtime/doc/diff.txt:
```diff
@@ -336,7 +337,129 @@ name or a part of a buffer name.  Examples:
+
+Example:
+
+Let's say we have the following files, side-by-side. We are interested in the
```
```suggestion
Let's say we have the following files, side-by-side.  We are interested in the
```

ychin replied:
Done

h-east commented on runtime/doc/options.txt:
```diff
@@ -3014,6 +3036,10 @@ A jump table for the options with a short description can be found at |Q_op|.
 
+		anchor		Anchor specific lines in each buffer to be
+				aligned with each other if 'diffanchors' is
+				set. See |diff-anchors|.
```
```suggestion
				set.  See |diff-anchors|.
```

ychin replied:
Done

h-east reviewed (commented):
There's a very minor issue here.

chrisbra commented on runtime/doc/diff.txt:
```diff
@@ -336,7 +337,129 @@ name or a part of a buffer name.  Examples:
+
+Marks: Set the |'a| mark on the `int foo()` lines in each file first before
+setting the anchors: >
+	set diffanchors='a
```
So if I understand correctly, this should be correct, right?

Note: By setting mark a in both buffers, we can set the global value of the 'diffanchors' option, when using a numerical address we must set the diffanchors option in each buffer to a different value.

ychin replied:
Yes, that's mostly correct, and it is what the example here is intended to communicate. (Technically for line numbers you can use a global option too, but it usually won't make sense to do so other than toy examples as the lines you want to anchor in each file would usually be different)

chrisbra commented on runtime/doc/options.txt:
```diff
@@ -2986,6 +2986,28 @@ A jump table for the options with a short description can be found at |Q_op|.
+	Each anchor line splits the buffer (the split happens above the
+	anchor), with each part being diff'ed separately before the final
+	result is joined.  When more than one {address} are provided, the
+	anchors will be sorted interally by line number.  If using buffer
```
```suggestion
	anchors will be sorted by line number.  If using buffer
```

chrisbra commented on runtime/doc/options.txt:
```diff
@@ -2986,6 +2986,28 @@ A jump table for the options with a short description can be found at |Q_op|.
+	local options, each buffer should have the same number of anchors
+	(extra anchors will be ignored).  This option is only used when
+	'diffopt' has "anchor" set.  See |diff-anchors| for more details and
+	examples.
```
Note: when using line-numbers as anchors, the diff hunks to be synced must be in the same order within each file.

Should we spell out this?

ychin replied:
> when using line-numbers as anchors

There's nothing special about using line numbers as anchors. It works the same way as marks and patterns. The diff anchor doesn't really care what you use at all. It just calls `get_address()` to translate a bunch of `{address}` to line numbers. That `{address}` could be `5`, or `'b`. It then just sort the results.

So basically, `set diffanchors=1,5,8` does the exact same thing as `set diffanchors=8,1,5`, which does the same thing as `set diffanchors=4-3,9-1,4+1` etc. They all get sorted internally to 1,5,8.

Is the documentation not clear on this? Maybe the part about "internally sorted by line number" could be confusing? The "line number" here refers to what each `{address}` resolves to, e.g. `'a` may resolve to line 201, and `'b` resolves to 560, and `45` resolves to… 45.

ychin replied:
I could reword this a bit if you think it's confusing.

chrisbra replied:
No that is fine. What I was trying to say, even so we can split the diff and force syncing on those pairs happens in the order in which they appear. It is not yet  possible to make the first pair in the first buffer sync with the second pair on the other buffer as those cannot be re-ordered. But it is fine for this PR and could be done later if it turns out to be a missing feature.

ychin replied:
Right. So what you mean is if we have `setlocal dia=3,6` on one buffer, and `setlocal dia=150,104` on another, you want to match lines 3/150, then 6/104 right? Instead of the current behavior which is to sort and match 3/104, 6/150.

I actually did think about this issue. I originally wanted to preserve the ordering as specified by the user, but when thinking more about it, the issue is that if you match lines 3/150, it will become impossible to match 6 with 104 afterwards, since line 6 (buf 1) is below the previous split, but line 104 (buf 2) is above the split. It means you have to ignore the second anchor which could be confusing in its own right.

So it's definitely doable, but it is trading one cons (sorting anchors leading to Vim ignoring the user-specified order) with another (not sorting anchors leading to some anchors needing to be dropped/ignored).

chrisbra commented on runtime/doc/diff.txt:
```diff
@@ -336,7 +337,129 @@ name or a part of a buffer name.  Examples:
+To use it, set anchors using 'diffanchors' which is a comma-separated list of
+{address} in each file, and then add "anchor" to 'diffopt'.  Internaly, Vim
+splits each file up into sections split by the anchors.  It performs the diff
+on each pair of sections separately before merging the results back.
```
I suppose the anchors are included in the pairs, right?

ychin replied:
The split happens *above* the anchor. You can think of the split as between two lines so we have to decide whether it happens above or below the anchor and it makes much more intuitive sense to do the split above. If we have an anchor at line 5, then the split will end up with two regions that we diff separately: (1,4), (5,$).

This is why the example in the docs for `:h diff-anchors` recommends using `set diffanchors='<,'>+1` if you want to anchor a visual range so the next split will happen *after* the last visually selected line.

I tried to communicate this in the `diffanchors` docs by saying the following, but I was struggling to word this without sounding too verbose and giving out too much unnecessary information:

>	Each anchor line splits the buffer (the split happens above the
>	anchor), with each part being diff'ed separately before the final
>	result is joined.

chrisbra reviewed (commented):
That is very nice. I think this can help reviewing changes much easier. Now we only need a way to tell which change on each side should be synced with which pair on the other diff side :)

ychin commented on src/ex_docmd.c:
```diff
@@ -4552,7 +4551,7 @@ get_address(
 		    else
 			curwin->w_cursor.col = 0;
 		    searchcmdlen = 0;
```
The `silent` flag is to make sure no side effects happen when you get `get_address()`. Previously it only made sure the pattern wouldn't be added to search history, but we need to also make sure it doesn't pollute `@/` (last search register), which is what `SEARCH_KEEP` does. Previously the only user for `silent = TRUE` is the `incsearch` code when it needs to parse an Ex command that user is typing, but it just manually worked around this by using `save_last_search_pattern()` to restore the now polluted `@/`. It's better to just fix this function.

See commit message from 2ebf6985d3c8 in this PR which describes this.

ychin committed:
Minor docs fix to use double space.
